### PR TITLE
feat: docker registry authentication

### LINKED ISSUE #798: Support Docker registry authentication
Currently we support only public Docker Hub images. We should define a config option for supporting authentication with Docker Hub, to later be expanded to custom registries after #797 is implemented.

Related to #1500

### PR DESCRIPTION
**Pull request by mxey**
I tried my hand at implementing authentication for Docker registries. This works for me with the Gitlab registry, and didn't break access to the public Docker Hub.

Since `config.endpoints` already exited internally, I added it to the config definitions so it would be merged with what's created by `config/env.js`.

I hardly speak JavaScript, so I'm wondering if this approach makes sense at all.

Closes #798

### CODE REVIEW

**Comment by rarkins** [comment]
@kayoub5 could you take a look at this and if it fits in with your plans for `endpoints`?

**Review by kayoub5 — APPROVED** [review]
Looks good to me

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/datasource/docker.js [hunk 1]
```diff
@@ -38,10 +40,18 @@ async function getAuthHeaders(registry, repository) {
 
     // prettier-ignore
     const authUrl = `${authenticateHeader.parms.realm}?service=${authenticateHeader.parms.service}&scope=repository:${repository}:pull`;
+    const { host } = URL.parse(registry);
+    const opts = endpoints.find({ platform: 'docker', host }, { json: true });
```
[the remark this answers is not in the document]
Comment (kayoub5) [reply lib/datasource/docker.js]: in `endpoints` the keys `platform` and `host` are treated as identifiers, for the endpoint, however they could be renamed to something more generic like `type` and `host`.

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/datasource/docker.js]: OK. I'll merge this first and get you to do the renaming later

**Review by rarkins — APPROVED** [review]

**Pull request merged**